### PR TITLE
[c10d][flight recorder] remove unintended assignment of entry

### PR DESCRIPTION
**Pull request by shuqiangzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119748

Summary:
auto& entry = entries_.at(*id % max_entries_);
entry = entries_.at(*id % max_entries_);
The above line of code has unintended consequence of invoking copy/assignment
of entry objects as ref itself cannot be re-assigned.

Also what could cause the crash is that the entry ref could become invalid if entries_ are
resized by other threads. and this could result in 'copy to a garbage
location'. The fix is to use a pointer which can be re-assigned after
re-acquiring the lock

Tests: python test/distributed/test_c10d_nccl.py NCCLTraceTest

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 1]
```diff
@@ -546,15 +546,15 @@ struct NCCLTraceBuffer {
 
     std::unique_lock<std::mutex> guard(mutex_);
 
-    auto& entry = entries_.at(*id % max_entries_);
-    if (entry.id_ == *id) {
-      update_state(entry);
+    Entry* entry = &entries_.at(*id % max_entries_);
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/TraceUtils.h:549]: Not sure if I missed anything, but it seems that the code would behave identically before and after the PR? https://godbolt.org/z/49Pn89aaG

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 2]
```diff
@@ -566,21 +566,21 @@ struct NCCLTraceBuffer {
       duration = getDurationFromFirstEvent(*startEvents, *endEvents);
       guard.lock();
 
-      // Refresh the entry ref, see if it has been overwritten
-      entry = entries_.at(*id % max_entries_);
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/TraceUtils.h:570]: @yifuwang the bug is in this line. the old code would invoke operator= and copy the requested entry contents into the value of the old reference, which could now belong to a different entry.  The new code would assign a new pointer value locally only, and let us check if it's still pointing to the same entry or not.  (it's a ring buffer)

**Review by wconstab — APPROVED** [review]
thanks for fixing my sad bug :(

**Comment by fegin** [comment]
@shuqiangzhang Thanks for the fix!

**Review by fegin — APPROVED** [review]
There seems to be a typo.

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 3]
```diff
@@ -528,15 +528,15 @@ struct NCCLTraceBuffer {
...
 
       if (compute_duration) {
-        can_compute_duration = entry.time_discovered_completed_.has_value() &&
-            entry.start_ && entry.end_;
-        startEvent = entry.start_;
-        endEvent = entry.end_;
+        can_compute_duration = entry->time_discovered_completed_.has_value() &&
+            entry->start_ && entry->end_;
+        startEvents = entry->start_;
+        endEvents = entry->end_;
```
[the remark this answers is not in the document]
Comment (shuqiangzhang) [reply torch/csrc/distributed/c10d/TraceUtils.h:539]: Merge conflicts.. Also seems pytorch trunk is broken

File: torch/csrc/distributed/c10d/TraceUtils.h (hunk 2 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/TraceUtils.h:552]: Ah I see it now! Thanks for highlighting

**Pull request closed without merging**